# Multiple Linear Regression Modeling for Sale Price Prediction at Elite Properties Group

# 1. Business Understanding


Elite Properties Group has the need to develop a reliable and accurate model that predicts the sale prices of residential properties based on their key attributes. By leveraging historical data, including features such as the number of bedrooms, bathrooms, square footage of living space, lot size, number of floors, property condition and year built, the agency aims to provide clients with precise pricing recommendations and maximize their returns on investment.

The primary objective of this project is to build a robust predictive model that accurately estimates the sale prices of residential properties listed by Elite Properties Group. By utilizing the available dataset, the agency aims to offer clients reliable and data-driven pricing advice, enhancing their confidence in the sales process and facilitating optimal pricing strategies.

# 2. Data Understanding
To gain a comprehensive understanding of the dataset provided by Elite Properties Group, I will examine the features and identify the key ones to use when creating models to predict the sale prices. The dataset `kc_house_data.csv` contains attributes such as: 
* Saleprice
* Number of Bedrooms
* Number of Bathrooms
* Year Built
* Date the house was sold
* Whether the house is on a waterfront
* Square footage of living space in the home
* Number of floors (levels) in house
* Quality of view from house

## Relevance of these features
The provided features in the dataset have significant relevance in predicting the sale prices of residential properties. The number of bedrooms and bathrooms provides insights into the property's size and functionality, which are crucial factors influencing its value. The square footage of the living space provides an indication of the overall spaciousness and potential use of the property, impacting its desirability and pricing. The number of floors reflects the property's layout and can affect its appeal to potential buyers.

The year built feature provides insights into the property's age, which can impact its condition, modernity, and potential maintenance costs. By considering these features collectively, we can develop a comprehensive model that accurately predicts sale prices and assists Elite Properties Group in providing valuable pricing advice to their clients.


# 3. Data Preparation
We first import the relevant packages for this analysis.

In [1]:
# importing the relevant packages
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
plt.style.use('seaborn-v0_8-whitegrid')

#ignoring warnings
import warnings
warnings.filterwarnings('ignore')

#enabling plotting of visualizations in the notebook
%matplotlib inline

## a. Previewing the dataset
We now load the dataset `kc_house_data` stored in the data folder. The data contains features that will be necessary for our modelling.

In [2]:
# loading the data using pandas and storing to variable df
df = pd.read_csv('data/kc_house_data.csv')

# previewing the first 10 rows
df.head(10)

id        date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014   221900.0         3       1.00         1180   
1  6414100192   12/9/2014   538000.0         3       2.25         2570   
2  5631500400   2/25/2015   180000.0         2       1.00          770   
3  2487200875   12/9/2014   604000.0         4       3.00         1960   
4  1954400510   2/18/2015   510000.0         3       2.00         1680   
5  7237550310   5/12/2014  1230000.0         4       4.50         5420   
6  1321400060   6/27/2014   257500.0         3       2.25         1715   
7  2008000270   1/15/2015   291850.0         3       1.50         1060   
8  2414600126   4/15/2015   229500.0         3       1.00         1780   
9  3793500160   3/12/2015   323000.0         3       2.50         1890   

   sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0      5650     1.0        NaN  NONE  ...      7 Average       1180   
1      7242     2.0         NO  NONE  ...      7 Average       2170   
2     10000     1.0         NO  NONE  ...  6 Low Average        770   
3      5000     1.0         NO  NONE  ...      7 Average       1050   
4      8080     1.0         NO  NONE  ...         8 Good       1680   
5    101930     1.0         NO  NONE  ...   11 Excellent       3890   
6      6819     2.0         NO  NONE  ...      7 Average       1715   
7      9711     1.0         NO   NaN  ...      7 Average       1060   
8      7470     1.0         NO  NONE  ...      7 Average       1050   
9      6560     2.0         NO  NONE  ...      7 Average       1890   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   
5         1530.0     2001           0.0    98053  47.6561 -122.005   
6              ?     1995           0.0    98003  47.3097 -122.327   
7            0.0     1963           0.0    98198  47.4095 -122.315   
8          730.0     1960           0.0    98146  47.5123 -122.337   
9            0.0     2003           0.0    98038  47.3684 -122.031   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  
5           4760      101930  
6           2238        6819  
7           1650        9711  
8           1780        8113  
9           2390        7570  

[10 rows x 21 columns]

## b. Removing features which are not necessary
Some of the columns in this dataset contain information which is not relevant for modelling. Before starting data cleaning, I'll remove some of these columns to remain with the most relevant.

Columns to be removed are: `date`, `view`, `sqft_above`, `sqft_basement`, `yr_renovated`, `zipcode`, `lat`, `long`, `sqft_living15` and `sqft_lot15`.

In [3]:
#dropping columns and assigning to new variable data
data = df.drop(columns = ['id','date','view','sqft_above','sqft_basement','yr_renovated','zipcode','lat','long','sqft_living15','sqft_lot15'], axis = 1)

#previewing the first 10 rows
data.head(10)

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors waterfront  \
0   221900.0         3       1.00         1180      5650     1.0        NaN   
1   538000.0         3       2.25         2570      7242     2.0         NO   
2   180000.0         2       1.00          770     10000     1.0         NO   
3   604000.0         4       3.00         1960      5000     1.0         NO   
4   510000.0         3       2.00         1680      8080     1.0         NO   
5  1230000.0         4       4.50         5420    101930     1.0         NO   
6   257500.0         3       2.25         1715      6819     2.0         NO   
7   291850.0         3       1.50         1060      9711     1.0         NO   
8   229500.0         3       1.00         1780      7470     1.0         NO   
9   323000.0         3       2.50         1890      6560     2.0         NO   

   condition          grade  yr_built  
0    Average      7 Average      1955  
1    Average      7 Average      1951  
2    Average  6 Low Average      1933  
3  Very Good      7 Average      1965  
4    Average         8 Good      1987  
5    Average   11 Excellent      2001  
6    Average      7 Average      1995  
7    Average      7 Average      1963  
8    Average      7 Average      1960  
9    Average      7 Average      2003

## c. Data Cleaning
Some of the columns may contain missing values,duplicates or incorrect formatting. In this stage, we are going to clean each column to remove these duplicates or missing values present. To start with, we'll first check for missing values in each table and deal with them.

## i) Identifying and handling missing values

In [4]:
#checking for missing values 
data.isnull().sum()

price             0
bedrooms          0
bathrooms         0
sqft_living       0
sqft_lot          0
floors            0
waterfront     2376
condition         0
grade             0
yr_built          0
dtype: int64

From the output above, the column `waterfront` contains 2376 missing values.

In [5]:
data['waterfront'].value_counts()

NO     19075
YES      146
Name: waterfront, dtype: int64

Since most of the values in column `waterfront` are No, I'll replace the missing values with NOs instead of just dropping the rows containing missing values.

In [6]:
#replacing missing values in waterfron with NO
data['waterfront'] = data['waterfront'].fillna('NO')

In [7]:
#checking again for missing values
data.isnull().sum()

price          0
bedrooms       0
bathrooms      0
sqft_living    0
sqft_lot       0
floors         0
waterfront     0
condition      0
grade          0
yr_built       0
dtype: int64

I've removed all rows containing missing values and now I check for duplicates.
## ii) Identifying and handling duplicates

In [8]:
#checking for duplicates
data.duplicated().value_counts()

False    21589
True         8
dtype: int64

The data contains three duplicate values. One needs to be dropped so that only unique values remain.
### Dropping duplicates

In [9]:
#dropping duplicates and keeping first
data = data.drop_duplicates(keep = 'first')

In [10]:
#checking again for duplicates
data.duplicated().value_counts()

False    21589
dtype: int64

Duplicates and missing values are now removed. Some of the columns are in a format that may make it difficult to create our models. Let's look at the data again:

In [11]:
data.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors waterfront  \
0  221900.0         3       1.00         1180      5650     1.0         NO   
1  538000.0         3       2.25         2570      7242     2.0         NO   
2  180000.0         2       1.00          770     10000     1.0         NO   
3  604000.0         4       3.00         1960      5000     1.0         NO   
4  510000.0         3       2.00         1680      8080     1.0         NO   

   condition          grade  yr_built  
0    Average      7 Average      1955  
1    Average      7 Average      1951  
2    Average  6 Low Average      1933  
3  Very Good      7 Average      1965  
4    Average         8 Good      1987

## iii) Converting format of columns
The column `bathrooms` and `floors` are decimals. Since we can't have a house with bathrooms in fractions or floors in fractions, we need to convert the type of these values to integer whole numbers.

In [12]:
# converting data types
data[['floors', 'bathrooms', 'price']] = data[['floors', 'bathrooms', 'price']].astype(int)

#previewing
data.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors waterfront  \
0  221900         3          1         1180      5650       1         NO   
1  538000         3          2         2570      7242       2         NO   
2  180000         2          1          770     10000       1         NO   
3  604000         4          3         1960      5000       1         NO   
4  510000         3          2         1680      8080       1         NO   

   condition          grade  yr_built  
0    Average      7 Average      1955  
1    Average      7 Average      1951  
2    Average  6 Low Average      1933  
3  Very Good      7 Average      1965  
4    Average         8 Good      1987

## iv) Dealing with categorical variables
Columns `waterfront`, `condition` and `grade` contain categorical variables. In order to begin modelling, we'll need to transform these variables into binary vector representations.
### Starting with `waterfront` column:

In [13]:
#checking the categories present
data['waterfront'].value_counts()

NO     21443
YES      146
Name: waterfront, dtype: int64

From above we can see that the column has only **YES** and **NO** values. I'll replace these values with a "1" and "0" respectively.
> 1 = YES

> 0 = NO

In [14]:
#replacing YES and NO with 1 and ) respectively
data['waterfront'] = data['waterfront'].replace(['YES','NO'], [1,0])

Previewing the data in column waterfront:

In [15]:
data['waterfront'].value_counts()

0    21443
1      146
Name: waterfront, dtype: int64

### Next, `condition` column.

In [16]:
#checking the categories present
data['condition'].value_counts()

Average      14013
Good          5676
Very Good     1701
Fair           170
Poor            29
Name: condition, dtype: int64

This column contains "Average", "Good", "Very Good", "Fair" and "Poor" values. Here we'll use ordinary encoding as the categories have a natural order or hierarchy. I'll replace as follows:
> 1 = Poor

> 2 = Fair

> 3 = Average

> 4 = Good

> 5 = Very Good

In [17]:
#replacing these values
data['condition'] = data['condition'].replace(['Poor','Fair', 'Average', 'Good', 'Very Good'], [1,2,3,4,5])

Previewing the data in this column:

In [18]:
data['condition'].value_counts()

3    14013
4     5676
5     1701
2      170
1       29
Name: condition, dtype: int64

### Finally, `grade` column.

In [19]:
#checking the categories present
data['grade'].value_counts()

7 Average        8974
8 Good           6061
9 Better         2611
6 Low Average    2038
10 Very Good     1134
11 Excellent      399
5 Fair            242
12 Luxury          89
4 Low              27
13 Mansion         13
3 Poor              1
Name: grade, dtype: int64

First, we need to identify the numeric portion of eachcategory as the values contain numeric values as well as string values. These are the numeric portions present with their respective assignment:
> 3 = Poor

> 4 = Low

> 5 = Fair

> 6 = Low Average

> 7 = Average

> 8 = Good

> 9 = Better

> 10 = Very Good

> 11 = Excellent

> 12 = Luxury

> 13 = Mansion

First of all, I'll need to remove any leading or trailing whitespaces from each value in the column to ensure consistent formatting. Then I will select only the integer values:

In [20]:
#removing whitespaces and selecting integer values
data['grade'] = data['grade'].str.split().str[0].astype(int)

Now that I've replaced all categorical variables with their respective binary representation, let't look at our data'

In [21]:
data.head(10)

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0   221900         3          1         1180      5650       1           0   
1   538000         3          2         2570      7242       2           0   
2   180000         2          1          770     10000       1           0   
3   604000         4          3         1960      5000       1           0   
4   510000         3          2         1680      8080       1           0   
5  1230000         4          4         5420    101930       1           0   
6   257500         3          2         1715      6819       2           0   
7   291850         3          1         1060      9711       1           0   
8   229500         3          1         1780      7470       1           0   
9   323000         3          2         1890      6560       2           0   

   condition  grade  yr_built  
0          3      7      1955  
1          3      7      1951  
2          3      6      1933  
3          5      7      1965  
4          3      8      1987  
5          3     11      2001  
6          3      7      1995  
7          3      7      1963  
8          3      7      1960  
9          3      7      2003

### Renaming `price` column to `SalePrice` for a better understanding:

In [22]:
#renaming column price
data.rename(columns = {'price': 'SalePrice'}, inplace = True)

In [23]:
#previewing the data
data.head()

SalePrice  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0     221900         3          1         1180      5650       1           0   
1     538000         3          2         2570      7242       2           0   
2     180000         2          1          770     10000       1           0   
3     604000         4          3         1960      5000       1           0   
4     510000         3          2         1680      8080       1           0   

   condition  grade  yr_built  
0          3      7      1955  
1          3      7      1951  
2          3      6      1933  
3          5      7      1965  
4          3      8      1987